include/lbann/transforms/transform.hpp

```diff
@@ -0,0 +1,111 @@
+////////////////////////////////////////////////////////////////////////////////
+// Copyright (c) 2014-2016, Lawrence Livermore National Security, LLC.
+// Produced at the Lawrence Livermore National Laboratory.
+// Written by the LBANN Research Team (B. Van Essen, et al.) listed in
+// the CONTRIBUTORS file. <lbann-dev@llnl.gov>
+//
+// LLNL-CODE-697807.
+// All rights reserved.
+//
+// This file is part of LBANN: Livermore Big Artificial Neural Network
+// Toolkit. For details, see http://software.llnl.gov/LBANN or
+// https://github.com/LLNL/LBANN.
+//
+// Licensed under the Apache License, Version 2.0 (the "Licensee"); you
+// may not use this file except in compliance with the License.  You may
+// obtain a copy of the License at:
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the license.
+////////////////////////////////////////////////////////////////////////////////
+
+#ifndef LBANN_TRANSFORMS_TRANSFORM_HPP_INCLUDED
+#define LBANN_TRANSFORMS_TRANSFORM_HPP_INCLUDED
+
+#include "lbann/base.hpp"
+#include "lbann/utils/description.hpp"
+#include "lbann/utils/random.hpp"
+#include "lbann/utils/type_erased_matrix.hpp"
+#include "lbann/utils/exception.hpp"
+
+namespace lbann {
+namespace transform {
+
+/**
+ * Abstract base class for transforms on data.
+ * 
+ * A transform takes a CPUMat and modifies it in-place. Transforms should
+ * be thread-safe, as one instance of a transform may be called concurrently
+ * within multiple threads.
+ *
+ * Because transforms may switch between underlying data types throughout the
+ * pipeline, everything is done in terms of a type_erased_matrix, which can
+ * swap between underlying data types.
+ */
+class transform {
+public:
+  transform() = default;
+  transform(const transform&) = default;
+  transform& operator=(const transform&) = default;
+  virtual ~transform() = default;
+
+  /** Create a copy of the transform instance. */
+  virtual transform* copy() const = 0;
+
+  /** Human-readable type name. */
+  virtual std::string get_type() const = 0;
+  /** Human-readable description. */
+  virtual description get_description() const {
+    return description(get_type() + " transform");
+  }
+
+  /** True if the transform supports non-in-place apply. */
+  virtual bool supports_non_inplace() const {
```

Comment:
I think the notion of "in-place" needs to be refined. We use it as a proxy for "converts `uint8_t` -> `DataType`", but that's rather non-obvious. We also allow different behaviors for `uint8_t` and `DataType` matrices. In particular, `uint8_t` matrices can change image dimensions while `DataType` matrices can't.

How about we make each transform either strictly in-place or out-of-place, so this function becomes `virtual bool is_inplace() const`. We also add functions that return the input and output matrix types, either as an enum or an `std::type_info`. Now, for example, `center_crop` is an out-of-place transform that operates on `uint8_t` matrices, which allows us to do things like zero padding without worrying about overflowing the buffer. The transform pipeline can also make sure that adjacent transforms are compatible.

There's still some messiness since we require that all `DataType` transforms are in-place, but I think it would work for now. In the future we can worry about using temporary `DataType` matrices and only putting the data reader's buffer at the final set of in-place transforms.

Reply:
> In particular, uint8_t matrices can change image dimensions while DataType matrices can't.

I don't think there's anything preventing that in the pipeline? We just don't have anything doing that right now.

These are all very good ideas. I think it best to defer them to a later PR in order to get the basic version of this done, however.

Reply:
The problem with changing the dimensions of `DataType` data is that it uses the actual output buffer for the input layer. You can't make it bigger since you'll go out of bounds and you can't make it smaller since the input layer will produce trash that the next layer needs to know to ignore.